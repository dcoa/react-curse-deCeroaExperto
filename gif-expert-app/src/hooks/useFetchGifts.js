import { useState, useEffect } from "react"
import { getGifs } from '../helpers/getGifts'
export const useFetchGifts = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
        getGifs(category)
            .then(img => setState({
                data: img,
                loading: false
            }))
        //El arreglo de dependencias esta vacio solo se ejecuta 1 unica vez
        //si cambia el valor de category vuelve y hace la petici√≥n (en este caso el valor no cambia)
    }, [category])

    return state;
}