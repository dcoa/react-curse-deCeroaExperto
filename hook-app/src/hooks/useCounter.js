import { useState } from "react"

export const useCounter = (initialState = 10) => {
    const [counter, setCounter] = useState(initialState);

    // el usuario ingresa el valor en el cual se incrementa el estado
    const increment = (factor) => {
        setCounter(counter + factor)
    };
    // el usuario ingresa el valor en el cual se decrementa el estado
    const decrement = (factor) => {
        setCounter(counter - factor)
    };
    // el usuario reestablece el contador
    const reset = () => {
        setCounter(initialState)
    };
    return {
        counter,
        reset,
        increment,
        decrement
    }
}
