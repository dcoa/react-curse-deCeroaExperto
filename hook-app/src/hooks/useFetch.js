import React, { useEffect, useState, useRef } from 'react'

export const useFetch = ( url ) => {
    const isMounted = useRef(true);
    const [state, setstate] = useState({data: null, loading: true, error: null})

         
    useEffect(() => {
        
        return () => {
            isMounted.current = false
        }
    }, [])

    useEffect(() => {
        // Realiza el fetch del componente que llama el use fetch solo si esta montado
        // limpiar el estado anterior antes de hacer el nuevo llamado al api
        setstate({data: null, loading: true, error: null})
        fetch( url )
        .then( resp => resp.json())
        .then(resp => {
            if( isMounted.current ) {

            setstate({error: null, loading: false, data: resp})
            }
        })
        .catch( () => setstate({data: null, loading: false, error: 'No se pudo cargar la info'}))
        

    }, [url])

    return state;
}
