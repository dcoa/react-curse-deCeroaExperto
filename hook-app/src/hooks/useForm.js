import { useState } from 'react'

//sirve para manejar los campos del formulario y leerlos
export const useForm = (initialState = {}) => {
   const [formValues, setformValues] = useState(initialState);
   const reset = () => setformValues(initialState)
   const handleInputChange = ({target}) => {
    setformValues({
        ...formValues,
        [target.name]: target.value
    })
}
return{
    formValues,
    handleInputChange,
    reset
}
}
