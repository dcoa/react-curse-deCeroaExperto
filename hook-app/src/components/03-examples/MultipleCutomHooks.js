import React from 'react'
import { useCounter } from '../../hooks/useCounter'
import { useFetch } from '../../hooks/useFetch'
import '../01-useState/counter.css'
export const MultipleCutomHooks = () => {
    const { counter, increment} = useCounter(1)
    const { loading, error, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`)
    //valida si la data existe el !! funciona para volver el null en false
    const {quote, author} = !!data && data[0];

    return (
        <div>
            <h1>Bracking Bad Quotes</h1>
            {loading ?
                <div className='alert alert-info text-center'>
                    Loading...
                </div>
                :
                <blockquote className='blockquote text-right'>
                    <p className='ab-0'>{quote}</p>
                    <footer className='blockquote-footer'>{author}</footer>
                </blockquote>
            }
            {
                !loading && <button className='btn btn-secondary' onClick={()=>increment(1)}>Next quote</button>
            }
        </div>

    )
}
