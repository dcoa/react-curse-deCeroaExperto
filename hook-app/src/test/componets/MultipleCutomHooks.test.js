import React from "react"
import { shallow } from "enzyme"
import { MultipleCutomHooks } from "../../components/03-examples/MultipleCutomHooks"
import { useFetch } from "../../hooks/useFetch"
import { useCounter } from "../../hooks/useCounter";

jest.mock("../../hooks/useCounter");
jest.mock("../../hooks/useFetch");

describe('Test on MultipleCutomHooks component', () => {

    beforeEach(()=>{
        useCounter.mockReturnValue({
            counter: 1,
        })
    })


    test('should render ', () => {
        // Crea un mock del use fetch para poder ver el comportamiento del componente
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })

        const wapper = shallow(<MultipleCutomHooks/>)
        expect(wapper).toMatchSnapshot();
    })
    
    test('should show the information about the quote', () => {
        useFetch.mockReturnValue({
            data: [{
                quote: 'Habia una vez',
                author: 'Jhon Doe'
            }],
            loading: false,
            error: null
        })

        const wapper = shallow(<MultipleCutomHooks/>)
        expect(wapper).toMatchSnapshot();
        expect(wapper.find('.alert').exists()).toBeFalsy();
        expect(wapper.find('p').text().trim()).toBe('Habia una vez');
        expect(wapper.find('footer').text().trim()).toBe('Jhon Doe');

    })
})
