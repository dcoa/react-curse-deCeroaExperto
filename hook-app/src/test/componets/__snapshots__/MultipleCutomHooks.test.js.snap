// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on MultipleCutomHooks component should render  1`] = `
<div>
  <h1>
    Bracking Bad Quotes
  </h1>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</div>
`;

exports[`Test on MultipleCutomHooks component should show the information about the quote 1`] = `
<div>
  <h1>
    Bracking Bad Quotes
  </h1>
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="ab-0"
    >
      Habia una vez
    </p>
    <footer
      className="blockquote-footer"
    >
      Jhon Doe
    </footer>
  </blockquote>
  <button
    className="btn btn-secondary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
